File: electron/recording-runtime-client.js
const fs = require('fs');
const http = require('http');
const https = require('https');
const path = require('path');

function validateRuntimeConnection(baseUrl, token) {
  const parsed = new URL(baseUrl);
  if (!['http:', 'https:'].includes(parsed.protocol) || parsed.username || parsed.password || !token) {
    throw new Error('A credentialed http or https Runtime profile is required.');
  }
  return parsed;
}

function privateSpeechEngine(config) {
  if (config?.speech?.providerId === 'private-runtime' && ['moss', 'funasr-meeting'].includes(config.speech.engineId)) return config.speech.engineId;
  // Compatibility for a task that was already written before the schema upgrade.
  if (config?.scheme === 'private-moss') return 'moss';
  if (config?.scheme === 'private-funasr') return 'funasr-meeting';
  throw new Error('A private Runtime speech selection is required.');
}

class RecordingRuntimeClient {
  constructor({ baseUrl, token, request = null }) {
    this.url = validateRuntimeConnection(baseUrl, token);
    this.token = token;
    this.request = request;
  }

  async health() { return this.json('GET', '/v1/health'); }
  async capabilities() { return this.json('GET', '/v1/capabilities'); }
  async getTask(taskId) { return this.json('GET', `/v1/tasks/${encodeURIComponent(taskId)}`); }
  async getResult(taskId) { return this.json('GET', `/v1/tasks/${encodeURIComponent(taskId)}/result`); }
  async cancel(taskId) { return this.json('POST', `/v1/tasks/${encodeURIComponent(taskId)}/cancel`); }
  async submit(type, payload) {
    if (!['translation', 'summary'].includes(type)) throw new Error('Speech tasks must use submitSpeech().');
    const { profileRevision, modelId, ...stagePayload } = payload;
    return this.json('POST', `/v1/${type}/tasks`, { profileRevision, ...(modelId ? { modelId } : {}), payload: stagePayload });
  }

  async submitSpeech(audioPath, profileRevision, config) {
    const source = await fs.promises.stat(audioPath);
    const boundary = `----sokuji-${Date.now().toString(16)}`;
    const fileName = path.basename(audioPath).replace(/"/g, '_');
    const fields = [
      ['profileRevision', profileRevision], ['engine', privateSpeechEngine(config)],
      ['sourceLanguageMode', config?.sourceLanguageMode || 'auto'],
      ['hotwords', JSON.stringify([...new Set((config?.hotwords || []).map((word) => String(word).trim()).filter(Boolean))])],
    ];
    if (config?.sourceLanguageMode === 'fixed' && config?.sourceLanguage) fields.push(['sourceLanguage', config.sourceLanguage]);
    const formFields = fields.map(([name, value]) => `--${boundary}\r\nContent-Disposition: form-data; name="${name}"\r\n\r\n${value}\r\n`).join('');
    const prefix = Buffer.from(`${formFields}--${boundary}\r\nContent-Disposition: form-data; name="audio"; filename="${fileName}"\r\nContent-Type: application/octet-stream\r\n\r\n`);
    const suffix = Buffer.from(`\r\n--${boundary}--\r\n`);
    return new Promise((resolve, reject) => {
      const transport = this.url.protocol === 'https:' ? https : http;
      const request = transport.request({
        protocol: this.url.protocol,
        hostname: this.url.hostname,
        port: this.url.port || undefined,
        path: `${this.url.pathname.replace(/\/$/, '')}/v1/speech/tasks`,
        method: 'POST',
        headers: {
          Authorization: `Bearer ${this.token}`,
          Accept: 'application/json',
          'Content-Type': `multipart/form-data; boundary=${boundary}`,
          'Content-Length': prefix.length + source.size + suffix.length,
        },
      }, (response) => this.collect(response, resolve, reject));
      request.on('error', (error) => reject(new Error(`Runtime speech upload failed: ${error.message}`)));
      request.write(prefix);
      const input = fs.createReadStream(audioPath);
      input.on('error', (error) => request.destroy(error));
      input.on('end', () => request.end(suffix));
      input.pipe(request, { end: false });
    });
  }

  json(method, requestPath, body) {
    const encoded = body === undefined ? null : Buffer.from(JSON.stringify(body));
    if (this.request) return this.request({ method, requestPath, body, token: this.token });
    return new Promise((resolve, reject) => {
      const transport = this.url.protocol === 'https:' ? https : http;
      const request = transport.request({
        protocol: this.url.protocol, hostname: this.url.hostname, port: this.url.port || undefined,
        path: `${this.url.pathname.replace(/\/$/, '')}${requestPath}`, method,
        headers: { Authorization: `Bearer ${this.token}`, Accept: 'application/json', ...(encoded ? { 'Content-Type': 'application/json', 'Content-Length': encoded.length } : {}) },
      }, (response) => this.collect(response, resolve, reject));
      request.on('error', (error) => reject(new Error(`Runtime request failed: ${error.message}`)));
      request.end(encoded || undefined);
    });
  }

  collect(response, resolve, reject) {
    const chunks = [];
    response.on('data', (chunk) => chunks.push(chunk));
    response.on('error', reject);
    response.on('end', () => {
      const text = Buffer.concat(chunks).toString('utf8');
      if (response.statusCode >= 400) return reject(new Error(`Runtime returned HTTP ${response.statusCode}: ${text.slice(0, 500)}`));
      try { resolve(JSON.parse(text)); } catch { reject(new Error('Runtime returned invalid JSON.')); }
    });
  }
}

module.exports = { RecordingRuntimeClient, privateSpeechEngine, validateRuntimeConnection };

File: sidecar/sokuji_sidecar/recording_runtime_client.py
"""Client for the isolated Linux Recording AI Runtime task API.

Secrets are supplied by the Electron main process to the calling engine only;
they are never part of a renderer message or a persisted recording job.
"""

import asyncio
import http.client
import json
from pathlib import Path
from dataclasses import dataclass
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen
from urllib.parse import urlsplit


class RuntimeClientError(RuntimeError):
    pass


@dataclass(frozen=True)
class RuntimeTaskClient:
    base_url: str
    bearer_token: str
    timeout_seconds: float = 30.0

    def __post_init__(self):
        if not self.base_url.startswith(("http://", "https://")):
            raise ValueError("Runtime base URL must use http or https")
        if not self.bearer_token:
            raise ValueError("Runtime bearer token is required")

    async def capabilities(self) -> dict[str, Any]:
        return await self._request("GET", "/v1/capabilities")

    async def submit(self, task_type: str, payload: dict[str, Any]) -> dict[str, Any]:
        if task_type not in {"translation", "summary"}:
            raise ValueError("Use the multipart speech uploader for speech tasks")
        return await self._request("POST", f"/v1/{task_type}/tasks", payload)

    async def submit_speech(self, audio_path: str, profile_revision: str) -> dict[str, Any]:
        return await asyncio.to_thread(self._submit_speech_sync, audio_path, profile_revision)

    async def get_task(self, task_id: str) -> dict[str, Any]:
        return await self._request("GET", f"/v1/tasks/{task_id}")

    async def get_result(self, task_id: str) -> dict[str, Any]:
        return await self._request("GET", f"/v1/tasks/{task_id}/result")

    async def cancel(self, task_id: str) -> dict[str, Any]:
        return await self._request("POST", f"/v1/tasks/{task_id}/cancel")

    async def _request(self, method: str, request_path: str, body: dict[str, Any] | None = None) -> dict[str, Any]:
        return await asyncio.to_thread(self._request_sync, method, request_path, body)

    def _request_sync(self, method: str, request_path: str, body: dict[str, Any] | None) -> dict[str, Any]:
        data = json.dumps(body).encode("utf-8") if body is not None else None
        request = Request(
            f"{self.base_url.rstrip('/')}{request_path}",
            data=data,
            method=method,
            headers={"Authorization": f"Bearer {self.bearer_token}", "Accept": "application/json", **({"Content-Type": "application/json"} if data else {})},
        )
        try:
            with urlopen(request, timeout=self.timeout_seconds) as response:
                return json.loads(response.read().decode("utf-8"))
        except HTTPError as error:
            detail = error.read().decode("utf-8", errors="replace")[:500]
            raise RuntimeClientError(f"Runtime request failed with HTTP {error.code}: {detail}") from error
        except URLError as error:
            raise RuntimeClientError(f"Runtime request could not be reached: {error.reason}") from error

    def _submit_speech_sync(self, audio_path: str, profile_revision: str) -> dict[str, Any]:
        source = Path(audio_path)
        if not source.is_file() or not profile_revision:
            raise ValueError("A readable audio path and profile revision are required")
        parsed = urlsplit(self.base_url)
        if parsed.query or parsed.fragment:
            raise ValueError("Runtime base URL must not contain a query or fragment")
        boundary = "----sokuji-runtime-upload"
        filename = source.name.replace('"', "_")
        prefix = (
            f"--{boundary}\r\nContent-Disposition: form-data; name=\"profileRevision\"\r\n\r\n{profile_revision}\r\n"
            f"--{boundary}\r\nContent-Disposition: form-data; name=\"audio\"; filename=\"{filename}\"\r\n"
            "Content-Type: application/octet-stream\r\n\r\n"
        ).encode("utf-8")
        suffix = f"\r\n--{boundary}--\r\n".encode("utf-8")
        connection_class = http.client.HTTPSConnection if parsed.scheme == "https" else http.client.HTTPConnection
        connection = connection_class(parsed.netloc, timeout=self.timeout_seconds)
        request_path = f"{parsed.path.rstrip('/')}/v1/speech/tasks" or "/v1/speech/tasks"
        try:
            connection.putrequest("POST", request_path)
            connection.putheader("Authorization", f"Bearer {self.bearer_token}")
            connection.putheader("Accept", "application/json")
            connection.putheader("Content-Type", f"multipart/form-data; boundary={boundary}")
            connection.putheader("Content-Length", str(len(prefix) + source.stat().st_size + len(suffix)))
            connection.endheaders()
            connection.send(prefix)
            with source.open("rb") as audio:
                while chunk := audio.read(1024 * 1024):
                    connection.send(chunk)
            connection.send(suffix)
            response = connection.getresponse()
            response_body = response.read().decode("utf-8", errors="replace")
            if response.status >= 400:
                raise RuntimeClientError(f"Runtime request failed with HTTP {response.status}: {response_body[:500]}")
            return json.loads(response_body)
        except OSError as error:
            raise RuntimeClientError(f"Runtime speech upload failed: {error}") from error
        finally:
            connection.close()
